internal/blockchain/ethereum/ethereum.go

```diff
@@ -172,67 +174,89 @@ func (e *Ethereum) Init(ctx context.Context, config config.Section, callbacks bl
 		}
 	}
 
-	if ethconnectConf.GetString(ffresty.HTTPConfigURL) == "" {
-		return i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "url", "blockchain.ethconnect")
+	if e.ethconnectConf.GetString(ffresty.HTTPConfigURL) == "" {
+		return i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "url", "blockchain.ethereum.ethconnect")
 	}
 
-	e.client = ffresty.New(e.ctx, ethconnectConf)
-
+	e.client = ffresty.New(e.ctx, e.ethconnectConf)
 	if fftmConf.GetString(ffresty.HTTPConfigURL) != "" {
 		e.fftmClient = ffresty.New(e.ctx, fftmConf)
 	}
 
-	e.instancePath = ethconnectConf.GetString(EthconnectConfigInstancePath)
-	if e.instancePath == "" {
-		return i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "instance", "blockchain.ethconnect")
+	e.topic = e.ethconnectConf.GetString(EthconnectConfigTopic)
+	if e.topic == "" {
+		return i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "topic", "blockchain.ethereum.ethconnect")
+	}
+	e.prefixShort = e.ethconnectConf.GetString(EthconnectPrefixShort)
+	e.prefixLong = e.ethconnectConf.GetString(EthconnectPrefixLong)
+
+	return nil
+}
+
+func (e *Ethereum) initStreams(contractIndex int) (err error) {
+
+	ctx := e.ctx
+	e.streams = &streamManager{client: e.client}
+	if e.contractConf.ArraySize() > 0 || contractIndex > 0 {
+		// New config (array of objects under "fireflyContract")
+		if contractIndex >= e.contractConf.ArraySize() {
+			return i18n.NewError(ctx, coremsgs.MsgInvalidFireFlyContractIndex, fmt.Sprintf("blockchain.ethereum.fireflyContract[%d]", contractIndex))
+		}
+		e.contractAddress = e.contractConf.ArrayEntry(contractIndex).GetString(FireFlyContractAddress)
```

Comment:
If I read it right, this code doesn't do anything to clean up the old subscriptions that are defunct. Not objecting to that, but making a note as I think it's significant (assuming I don't find that elsewhere).

Reply:
I might have missed it, but I'm looking out for the code that records the conclusive `endBlock` for the contract index `0`. I see us calling `bi.Start()` which calls into this code, but I don't see anything that prevents further events being consumed.

Otherwise, in the middle of the processing of the `OperatorActionMigrate` message, we have delete the old stream in a way that prevents any more events possibly being processed. I don't know that we have that API contract with blockchain event listeners, that they have to be able to process a REST API to delete a stream in the middle of an ack.

Reply:
Good point. I was actually planning to delete and recreate the stream and then forgot that piece. But your suggestion of an end block is probably better, if I can figure out where to record it...

Reply:
OK, made a significant additional commit to handle these two changes:
1. The "final event" of every previous FireFly contract is stored forever, and events from that contract past that point are ignored
2. When migrating to a new contract, the new subscription is simply added to the existing stream (no attempt to clean up old ones and no need to restart the websocket)

Reply:
I also want to note that with this model we never delete the old subscriptions (just ignore any events they produce). That leaves a bit of unnecessary overhead on ethconnect, but it was just difficult enough that I didn't want to mess with it at this time (as noted above, deleting a subscription while in the process of processing/acking an event _from_ that subscription seems like it could cause problems).